drivers: flash: Add Ambiq flash controller driver

aaronyegx opened the pull request:
This PR creates the flash controller driver for Ambiq Apollo4 SoCs.
It also instantiates the flash controller for Apollo4 Blue Plus SoC and enable the flash controller for apollo4p_blue_kxr_evb.
It mainly targets to do data storage for the Bluetooth applications which need Zephyr Settings subsystem at this moment.

Test passed on the following cases:
--samples/drivers/flash_shell
--tests/drivers/flash/common

aaronyegx committed:
dts: arm: ambiq: Add flash controller instance to Apollo4 Blue Plus SoC.

This commit instantiates the flash controller.

Signed-off-by: Aaron Ye <aye@ambiq.com>

Laczen commented:
Hi @aaronyegx this generally looks ok, some possible improvements:
1. Also allow for non-multithreading by removing the k_sem mutex and the k_sem_take and k_sem_get routines in that case,
2. There is no check on the alignment of the write data address and length to the write-block-size,
3. There is no check on the alignment of the erase start and length to the erase-block-size,

Since the erase is a "simulated" erase you can set the erase-block-size to whatever is wanted, but from the code it seems that it needs to be a multiple of 4 byte. Also the write-block-size seems to be 4 byte and there is no need to set it to 16 byte.

There might also be a problem when the data passed to the write function would not be aligned to 4 byte. This is easily checked by creating a const char tst="TESTDATA" and then trying to write 4 byte starting from tst[0], tst[1], ... To remedy from this you can use the UNALIGNED_GET() macro.

aaronyegx committed:
boards: arm: apollo4p_blue_kxr_evb: Enable flash controller.

This commit enables flash controller instance for apollo4p_blue_kxr_evb.

Signed-off-by: Aaron Ye <aye@ambiq.com>

aaronyegx commented:
> Hi @aaronyegx this generally looks ok, some possible improvements:
> 
> 1. Also allow for non-multithreading by removing the k_sem mutex and the k_sem_take and k_sem_get routines in that case,
> 2. There is no check on the alignment of the write data address and length to the write-block-size,
> 3. There is no check on the alignment of the erase start and length to the erase-block-size,
> 
> Since the erase is a "simulated" erase you can set the erase-block-size to whatever is wanted, but from the code it seems that it needs to be a multiple of 4 byte. Also the write-block-size seems to be 4 byte and there is no need to set it to 16 byte.
> 
> There might also be a problem when the data passed to the write function would not be aligned to 4 byte. This is easily checked by creating a const char tst="TESTDATA" and then trying to write 4 byte starting from tst[0], tst[1], ... To remedy from this you can use the UNALIGNED_GET() macro.

Hi @Laczen , thanks for your review. I updated the commits as you commented:
1. Included the k_sem mutex and related k_sem_take/k_sem_give in CONFIG_MULTITHREADING, so it should support multithreading and non-multithreading scenarios.
2. Actually, there is check on the Ambiq HAL APIs but I also added the address and length alignment pre-check for flash write in the flash controller driver to make it clear.
3. Added address and length check for flash erase as well.

The Ambiq Apollo4x flash write and erase operation require 4-word alignment, so the write-block-size is configured to 16 bytes. And the erase-block-size is configured to 2048 bytes (which seems popular in existing boards). Some low size (e.g., 16 bytes) may get EDEADLK error in settings_nvs_backend_init() when it is used for Bluetooth settings.

Could you talk another look on the changes? Thank you.

Laczen commented:
@aaronyegx thank you for the updates.

Regarding the multithreading support, the code can be cleaned up a little by defining at the top of the code:
```
#if defined(CONFIG_MULTITHREADING)
static struct k_sem flash_ambiq_sem;
#define FLASH_SEM_INIT() k_sem_init(&flash_ambiq_sem, 1, 1)
#define FLASH_SEM_TAKE() k_sem_take(&flash_ambiq_sem, K_FOREVER)
#define FLASH_SEM_GIVE() k_sem_give(&flash_ambiq_sem)
#else
#define FLASH_SEM_INIT()
#define FLASH_SEM_TAKE()
#define FLASH_SEM_GIVE()
#endif
```
and then using `FLASH_SEM_INIT()`, `FLASH_SEM_TAKE()` and `FLASH_SEM_GIVE()` in the code. The `flash_data` variable can then be dropped.

Regarding the write alignment, there is no need to double check, add a note somewhere in the code that this alignment is checked in the hal.

Regarding the erase alignment, there is no requirement to check the alignment to the erase-block-size. The subsystem that uses the flash and is dependent on the block size will use the page layout to get the block size and issue a correct erase, while a subsystem that does not use it can issue a smaller erase when needed (@de-nordic, this might be a good method to avoid defining a flash that has "no-erase": a read of an area could be done to determine that an area is erased and if not a smaller erase could be performed).

Have you checked writing unaligned data ? Taking into account the 16 byte write block size you should define a 20 byte array of test data: const char tst[16]="TST0TST1TST2TST3TST4" and writing  should be succesfull for &tst[0], &tst[1], &tst[2], &tst[3].

The choice of 2048 bytes for the erase-block-size is good.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,175 @@
+	return ret;
+}
+
+static int flash_ambiq_erase(const struct device *dev, off_t offset, size_t len)
```
MRAM does not require erase and does not provide it.
Emulated erase, in case of any xRAM devices, reduces use count by 1 the same as write does. In case of flash, the reduction only happens at erase operation, write does not affect wear.

de-nordic commented:
> (@de-nordic, this might be a good method to avoid defining a flash that has "no-erase": a read of an area could be done to determine that an area is erased and if not a smaller erase could be performed).

It kills device twice as fast as flash device with the same  bit flip count;  write is destructive operation in MRAM and it is done twice, while in flash erase is destructive, write is not. So write/write != erase/write in wear counter.

de-nordic commented:
> > (@de-nordic, this might be a good method to avoid defining a flash that has "no-erase": a read of an area could be done to determine that an area is erased and if not a smaller erase could be performed).
> 
> It kills device twice as fast as flash device with the same bit flip count; write is destructive operation in MRAM and it is done twice, while in flash erase is destructive, write is not. So write/write != erase/write in wear counter.

This is not flash device, and we name it flash anyway, so maybe no-erase is not that bad idea at all, as naming does not seem to be so important anyway and we are willing to give up on device reliability just to pretend it is flash.

There will be issue reported further or later that MRAM device are dying faster than expected, no-erase flag at least would allow us to have intermediate solution in the entire system before we get into right one.

Laczen commented:
@de-nordic, I agree that this is not a flash device but ...

1. It follows all the requirements set by the flash api,
2. MRAM cycles seem to be a lot higher than flash cycles, so erasing them before write when needed (in case of e.g. nvs) might not be as problematic as for flash,

Regarding the no-erase flag could we take this out of this PR. I am sure this can be resolved.

aaronyegx commented:
Hi @Laczen , yes there was unalignment writing issue in previous implementation. It failed in your suggested testing. I use UNALIGNED_GET() to make it work now. And clean up the multithreading support codes. Thanks so much.

aaronyegx commented:
Hi @de-nordic , thanks for your review. Yes Ambiq Apollo4x SoC has no flash and we always use MRAM for the storage purpose, e.g., Bluetooth NVM. Many Zephyr Bluetooth samples are using Settings subsystem for the data storage then the flash controller driver should be supported basically.

CONFIG_BT_SETTINGS=y
CONFIG_FLASH=y
CONFIG_FLASH_PAGE_LAYOUT=y
CONFIG_FLASH_MAP=y
CONFIG_NVS=y
CONFIG_SETTINGS=y

So, we consider MRAM is "flash" at this case. The erase operation is not necessary for MRAM writing but I would like to keep the API for some necessary scenarios. Do you think it is acceptable?

Laczen commented:
> Hi @Laczen , yes there was unalignment writing issue in previous implementation. It failed in your suggested testing. I use UNALIGNED_GET() to make it work now. And clean up the multithreading support codes. Thanks so much.

Some nitpicks from my side:
1. Now that you have split up the write in a loop it is no longer possible to count on the hal to do the checking of alignment, so I think you previous change that did the alignment check with write-block-size needs to be inserted. Now the loop would do a part of a write and fail at the end. **Don't forget to remove the remark.**
2. Don't use the numbers 16 and 4 in code (in the write or alignment check):
    16 should be replaced by `FLASH_WRITE_BLOCK_SIZE`,
    4 (in the write routine) should be replaced by `FLASH_WRITE_BLOCK_SIZE/sizeof(uint32_t)`.

de-nordic commented:
> Hi @de-nordic , thanks for your review. Yes Ambiq Apollo4x SoC has no flash and we always use MRAM for the storage purpose, e.g., Bluetooth NVM. Many Zephyr Bluetooth samples are using Settings subsystem for the data storage then the flash controller driver should be supported basically.
> 
> CONFIG_BT_SETTINGS=y CONFIG_FLASH=y CONFIG_FLASH_PAGE_LAYOUT=y CONFIG_FLASH_MAP=y CONFIG_NVS=y CONFIG_SETTINGS=y
> 
> So, we consider MRAM is "flash" at this case. The erase operation is not necessary for MRAM writing but I would like to keep the API for some necessary scenarios. Do you think it is acceptable?

Can we at least have it define new compatibles like "soc-nvm-mram" and "ambiq,mram-controller" or "ambiq,nvm-controller".
This way, in the future, when the MRAM driver is actually provided DTS compatibles will be the same.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,182 @@
+#endif /* CONFIG_MULTITHREADING */
+
+static const struct flash_parameters flash_ambiq_parameters = {
+	.write_block_size = FLASH_WRITE_BLOCK_SIZE,
```
I think assignments like this should be `MAX(hardware_specific, DTS value)`. And there probably should be build error when DTS is smaller than hardware minimal and not a multiply of it?

aaronyegx replied:
I don't clearly understand this point. The write_block_size parameter here is set by the DT_PROP(SOC_NV_FLASH_NODE, write_block_size) which should be configured properly in DTS. Are you saying we need to define a macro for the hardware_specific requirement (16 for Ambiq Apollo4x)?

#define HW_WRITE_BLOCK_SIZE  16
.write_block_size = MAX(HW_WRITE_BLOCK_SIZE, FLASH_WRITE_BLOCK_SIZE)

Laczen replied:
I also don't really understand this point. The write-block-size parameter is a real hardware device property and the value in the dts should be correct.

For the erase-block-size (this is a "make-compatible" value and could be adapted to fit the needs) I could understand that it is checked as being a multiple of write-block-size.

de-nordic replied:
My point is simple here: the device has design constrains on the write block size that should be maintained by the driver at compilation phase. The user may override write block size in DTS with value that still is correct and will work, like multiple of the hardware characteristic of the write-block-size.
That is why the write-block-size taken from the DTS should be checked against what the driver will actually accept or could work with.
I know that there are drivers that basically ignore the DTS and have the value encoded in driver only, so the DTS is more informative than actually used for anything, but here it has been decided to take the values as something that is used to compile the driver, so the driver should have compile time checks on the user input.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,182 @@
+
+static bool flash_ambiq_valid_range(off_t offset, size_t len)
+{
+	if ((offset < 0) || (offset + len) > SOC_NV_FLASH_SIZE) {
```
`offset < 0 || offset >=  SOC_NV_FLASH_SIZE || (SOC_NV_FLASH_SIZE - offset) < len` should catch all the int wraparounds.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,182 @@
+
+	int ret = 0;
+	uint32_t critical = 0;
+	uint32_t aligned[4] = {0, 0, 0, 0};
```
`= { 0 };`  works as good

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,182 @@
+
+	FLASH_SEM_TAKE();
+
+	erase_word = ((uint32_t)(flash_ambiq_parameters.erase_value << 24)) |
+		     ((uint32_t)(flash_ambiq_parameters.erase_value << 16)) |
+		     ((uint32_t)(flash_ambiq_parameters.erase_value << 8)) |
+		     ((uint32_t)flash_ambiq_parameters.erase_value);
```
Thad does not need to be within SEM.
Also, unless you are supporting multiple instances with different erase_values, the 32bit erase value may be pre-calculated for this module at compile time.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,182 @@
+
+#if CONFIG_FLASH_PAGE_LAYOUT
+static const struct flash_pages_layout pages_layout = {
+	.pages_count = DT_REG_SIZE(SOC_NV_FLASH_NODE) / FLASH_ERASE_BLOCK_SIZE,
```
`SOC_NV_FLASH_SIZE` is already defined for `DT_REG_SIZE(SOC_NV_FLASH_NODE)`.

de-nordic commented on drivers/flash/Kconfig.ambiq:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config FLASH_AMBIQ
+	bool "Ambiq flash driver"
```
Can we call it `Amibq flash on MRAM driver" or something like that?

de-nordic commented:
> > Can we at least have it define new compatibles like "soc-nvm-mram" and "ambiq,mram-controller" or "ambiq,nvm-controller". This way, in the future, when the MRAM driver is actually provided DTS compatibles will be the same.
> 
> Hi @de-nordic , I think it should be OK to define new compatible "soc-nv-mram". Current "soc-nv-flash" defines "erase-block-size" and "write-block-size" properties, do you think the "soc-nv-mram" should define the same?
> 
> But for name of API and c file, should we still use "flash_" as the prefix?

Definitions here are local/ static and completely unique for the device, so you can call them as you wish, even with prefix ambiq_nvm_; think what would allow you to identify the callback, in the easiest way, when you look at stack trace; or allow you to find property by identifier.

The write block size is needed by the characteristic of the device; the erase block size should be defined as optional, but is needed now by the subsystem.

If we are able to bring in proper mram support, with the dts binding done right we should be able to reuse  most of the work provided by the pr, without significantly altering dts afterwards.

Laczen commented:
I think the compatible should remain "soc-nv-flash". As a "flash" device the write-block-size needs to be defined. The erase-block-size could be an optional dts value but it could also be fixed in the driver and reflected in the page layout. IMO it is good as proposed.

If the compatible name is changed partition definitions will fail to be created.

The config entry would indeed better be changed to "Ambiq flash driver on mram".

aaronyegx commented:
Hi @de-nordic @Laczen , thanks for your continued support on this PR. I address most of the comments you put in the code.
For the naming, as @Laczen commented, I also prefer to keep "soc-nv-flash", "flash-controller" to align the current "flash" driver naming. Using "soc-nv-mram" or "nvm-controller" may bring some conflict or confusion to the current framework. For example, we can only use "zephyr, flash-controller" as the "chosen" for this instance or we will get building or run failure in the Zephyr examples or test cases, e.g., the samples/drivers/flash_shell. 
I think supporting MRAM in Zephyr can be another topic which should be discussed more outside of this PR.
I modify the description of flash driver configuration to "Ambiq flash driver on MRAM", it should make sense.

aaronyegx commented:
Hi @de-nordic , could you take another look at the current implementation? Thanks.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,190 @@
+
+static bool flash_ambiq_valid_range(off_t offset, size_t len, bool write)
+{
+	if ((offset < 0) || offset >= SOC_NV_FLASH_SIZE || (SOC_NV_FLASH_SIZE - offset) < len) {
```
I think that the last element of the expression' (SOC_NV_FLASH_SIZE - offset) < len)' , should be  `(SOC_NV_FLASH_SIZE - offset) > len)`, `len` needs to fit in the size that is left from the `SOC_NV_FLASH_SIZE` when we start the op at the `offset`.

Laczen replied:
IMO the code is correct, if the available space is less than  `len` an error should be generated.

de-nordic replied:
Ah right, this is condition for error not pass.

de-nordic commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,190 @@
+		return false;
+	}
+
+	if (write) {
+		/* write address must be block size aligned and the write length must be multiple of
+		 * block size.
+		 */
+		if (((uint32_t)offset & (FLASH_WRITE_BLOCK_SIZE - 1)) ||
+		    (len & (FLASH_WRITE_BLOCK_SIZE - 1))) {
+			return false;
+		}
+	}
```
Your driver, your choice: this part of code is valid for only one call in the unit here, yet the `if (write)` will be executed for every call of the function.
This part should, IMHO, be moved to the `flash_ambiq_write` as it has usage only there.

Laczen commented:
@de-nordic, regarding the checking of the `write-block-size`, as more and more the dts entries are used as configuration entries this raises the question if these values should be considered "configuration" or "definition".
I don't mind it being checked but we should have a clear idea on how to treat the dts values. If they are all considered "configuration" all drivers should check them and use them if specified. If they are considered "definitions" they could just be used. If some of them are considered "definitions" while others are considered "configurations" the proposed PR treats them as "definitions" and this would be a valid choice.

aaronyegx commented:
Hi @de-nordic , @Laczen , regarding the checking of the write-block-size, can we go with the below implementation?

#define MIN_WRITE_SIZE         16
#define FLASH_WRITE_BLOCK_SIZE MAX(DT_PROP(SOC_NV_FLASH_NODE, write_block_size), MIN_WRITE_SIZE)

#if (FLASH_WRITE_BLOCK_SIZE & (MIN_WRITE_SIZE - 1) != 0)
#error "The flash write block size must be a multiple of 16!"
#endif

Laczen commented:
> Hi @de-nordic , @Laczen , regarding the checking of the write-block-size, can we go with the below implementation?
> 
> #define MIN_WRITE_SIZE 16 #define FLASH_WRITE_BLOCK_SIZE MAX(DT_PROP(SOC_NV_FLASH_NODE, write_block_size), MIN_WRITE_SIZE)
> 
> #if (FLASH_WRITE_BLOCK_SIZE & (MIN_WRITE_SIZE - 1) != 0) #error "The flash write block size must be a multiple of 16!" #endif

No problem with adding the check for me, the error reporting should be done using BUILD_ASSERT().

aaronyegx commented:
> > Hi @de-nordic , @Laczen , regarding the checking of the write-block-size, can we go with the below implementation?
> > #define MIN_WRITE_SIZE 16 #define FLASH_WRITE_BLOCK_SIZE MAX(DT_PROP(SOC_NV_FLASH_NODE, write_block_size), MIN_WRITE_SIZE)
> > #if (FLASH_WRITE_BLOCK_SIZE & (MIN_WRITE_SIZE - 1) != 0) #error "The flash write block size must be a multiple of 16!" #endif
> 
> No problem with adding the check for me, the error reporting should be done using BUILD_ASSERT().

Hi @de-nordic , could you take another look at this? Thank you.

Laczen commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,192 @@
+	 */
+	if (((uint32_t)offset & (FLASH_WRITE_BLOCK_SIZE - 1)) ||
+	    (len & (FLASH_WRITE_BLOCK_SIZE - 1))) {
+		return false;
```
I think this should be `-EINVAL`, you can combine this with the above condition.

Laczen commented on drivers/flash/flash_ambiq.c:
```diff
@@ -0,0 +1,192 @@
+	if (!flash_ambiq_valid_range(offset, len)) {
+		return -EINVAL;
+	}
+
```
The alignment check should also be added here.

aaronyegx replied:
@Laczen I added alignment check here in the previous implementation, but @de-nordic commented we should put it to flash_write API since only flash_write needs alignment check, so I modified it. How do you think about it?

aaronyegx replied:
For erase operation, there is address alignment and length check in the HAL API, so I don't think we need to add it. I can leave a comment in the code.

Laczen replied:
Yes, in that case it is not needed. As this is one hal operation we can rely on the check by the hal (I forgot about this). You can disregard this remark.

A little comment line would be welcome for future users to understand why a check is not needed.

Laczen reviewed (changes requested):
I have spotted some minor things that need to be changed.

aaronyegx committed:
drivers: flash: Add Ambiq flash controller driver.

This commit adds flash controller driver for Ambiq Apollo4 SoCs.

Signed-off-by: Aaron Ye <aye@ambiq.com>

Laczen reviewed (approved):
LGTM, thanks for the changes and the work you have put into this.

aaronyegx commented:
Hi @de-nordic , could you help revisit and move forward this PR? Thanks.
